trade-finance-manager-ui/server/controllers/login/index.js

```diff
@@ -1,53 +1,118 @@
+    const errorMessage = `Detected login redirect issue - please wait for ${SSO.REDIRECT_COUNTER.TIME_PERIOD/1000} seconds or contact us using details bellow.`;
+    return res.status(500).render('_partials/problem-with-service.njk', { error: { message: errorMessage }});
   }
 
-  if (loginErrors.length) {
-    return res.render('login.njk', {
-      errors: validationErrorHandler(loginErrors),
-    });
+  // Track redirects
+  res.cookie(SSO.REDIRECT_COUNTER.COOKIE_NAME, ssoRedirectNo, { maxAge: SSO.REDIRECT_COUNTER.TIME_PERIOD, httpOnly: true, sameSite: 'none', secure: true });
+
+  const apiResponse = await api.getAuthLoginUrl();
+  if (apiResponse.loginUrl) {
+    req.session.auth = apiResponse;
+    return res.redirect(apiResponse.loginUrl);
   }
 
-  return res.redirect('/home');
+  console.error('TFM-UI - Login url generation failed. TFM-API response: %O', apiResponse);
+  const errorMessage = `Login url generation failed - please try later or contact us using details bellow`;
+  return res.status(500).render('_partials/problem-with-service.njk', { error: { message: errorMessage } });
 };
 
-const logout = (req, res) => {
-  req.session.destroy(() => {
-    res.redirect('/');
-  });
+/**
+ * handleSsoRedirect
+ * Because of security restrictions we need to handle 2 POST calls in same URL.
+ * 1) Step1 - POST is from SSO Authority we don't have session cookie and will prepare local form that will autosubmit.
+ * 2) Step2 - POST is from our form and now we have session cookie and can verify SSO authority data.
+ * 3) Send SSO authority data to TFM-API for verification and login.
+ * 4) Redirect user to original location, most likely /.
+ * @param {Object} req: Express request
+ * @param {Object} res: Express response
+ * @returns {Promise<Object>} Express response with redirect or error.
+ */
+const handleSsoRedirect = async (req, res) => {
+  try {
+    const { body, session } = req;
+
+    if (!body.formId) {
+      return loginService.acceptExternalSsoPost(req, res);
+    }
+    if (!session?.auth?.pkceCodes) {
+      console.error('SSO login session details are missing. Start login from beginning.');
+      return res.redirect('/');
+    }
+
+    const {
+      auth: {
+        pkceCodes,
+        authCodeUrlRequest,
+        authCodeRequest,
+      },
+    } = session;
+    const { code, state } = body;
+
+    const apiResponse = await api.processSsoRedirect({ pkceCodes, authCodeUrlRequest, authCodeRequest, code, state });
+
+    if (apiResponse.token) {
+      req.session.userToken = apiResponse.token;
+      req.session.user = apiResponse.tfmUser;
+
+      // Unset data used for SSO validation
+      delete req.session.auth;
+
+      // Unset cookie used for SSO redirect counting
+      res.clearCookie(SSO.REDIRECT_COUNTER.COOKIE_NAME);
+
+      return res.redirect(apiResponse.redirectUrl);
+    }
+    console.error('TFM-UI - login failed in TFM-API, redirect to /');
+    return res.redirect('/');
+  } catch(error) {
```

Comment:
Sorry I know I suggested this, future me seems to dislike past me...

Reply:
It should be fine because all api calls, like `api.processSsoRedirect`, have they own try catch block and don't expose secrets/errors. 

I can't see sensitive error message even if I try.